### ISSUE TITLE
When a node rerenders due to new context api update, its siblings rerender

### ISSUE THREAD

**Issue by macrozone** [facebook/react#13086]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

Nodes that use new context api force siblings nodes to rerender

**What is the expected behavior?**

Siblings should not rererender when a node updates due to context updates

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

16.3.1

Background:

i use `react-navigation` with react-native. react-navigation uses the new context-api. `withNavigationFocus`-HOC wraps the component with a `NavigationConsumer`, a context consumer.

In this example i get unexpected rerenders:

```
const SomethingThatHasNavigation = withNavigationFocus(
  class extends React.Component {
    render() { 
      // i should rerender whenever navigation changes. That's fine. Its value is from context
      return null
    }
  },
)

const ShouldNotRerender = class extends React.Component {
  render() {
    console.log('i should not rerender when my sibiling updates')
    return null
  }
}

const Test = class extends React.PureComponent {
  render() {
    console.log('test renders once')
    return (
      <View>
        <SomethingThatHasNavigation />
        <ShouldNotRerender />
      </View>
    )
  }
}
```

What's wrong? is this by design?

**Comment by aweary** [comment]
@macrozone can you please test with the latest (16.4.1) release?

**Comment by danburzo** [comment]
I'm not sure I understand the whole setup, but I'll try to answer yo the best of my knowledge.

Whenever a Context Provider re-renders, all its descendants re-render, as would happen with a normal component. What the context API does is that, in addition to the normal rules of re-rendering (governed by all the `shouldComponentUpdate` methods down the component tree), is that it _forces_ updates to all its Consumers, regardless of whether they're in a pure component that would normally bail out of the update. If View is a consumer, all its children will re-render. 

I hope I'm not too off with the answer!

**Comment by macrozone** [comment]
@danburzo that's how i understand it: if a component rerenders due to props, state or context changes, all its descendants rerender unless they `shouldComponentUpdate` or PureComponent. But in my setup, a _sibling_ rerenders that should not be affected at all. The parent of both siblings does not rerender as expected. So the full chain is

Provider --> Parent --> A, B (siblings)

A is the component that should not update, B is Consumer.

If context changes i would expect only B to update, but A updates as well, while Parent in my case does not update (it has shouldComponentUpdate return false).

I have to check whether i can setup a simple reproduction, but i wanted to ask, if my scenario could happen in react 16+

@aweary its in a react-native project, can i simple use react 16.4.1?

**Comment by macrozone** [comment]
@aweary  i could install 16.4.1, but with the same result :-(

**Comment by danburzo** [comment]
@macrozone you are correct in your assumption. In fact, this behaves exactly like that:

```jsx
const ctx = React.createContext();

class App extends React.Component {
  
  state = {
    count: this.props.count
  }
  
  incrementCount = e => {
    this.setState(current_state => ({ count: current_state.count + 1 }));
  }

  render() {
    return (
      <div onClick={ this.incrementCount }>
        <ctx.Provider value={this.state.count}>
          <PureParent/>
        </ctx.Provider>
      </div>
    );
  }
}

class PureParent extends React.PureComponent {
  render() {
    return (<React.Fragment>
      <ConsumerChild/>
      <OtherSibling/>
    </React.Fragment>);
  }
}

const ConsumerChild = props => <ctx.Consumer>{ count => <p>{'count: ' + count}</p> }</ctx.Consumer>;

let render_count = 1;
const OtherSibling = props => <p>Other sibling: {render_count++}</p>

ReactDOM.render(<App count={0}/>, document.querySelector('#app'));
```

[See it on CodePen](https://codepen.io/danburzo/pen/XYqbNy)

So my guess is that it's something related to your react-navigation setup?

**Comment by gaearon** [comment]
@macrozone Can you please provide a complete example using React 16.4.1 on CodePen? Thanks.

**Comment by macrozone** [comment]
@danburzo thank you so much for your help!

This codepen works exactly as i would expect, so it has to be something else. I'll try to reproduce on react-native and would come back if i have further insights

**Comment by gaearon** [comment]
Note RN might include a bug that was fixed in 16.4.0. Just updating React package won’t help you in this case: the reconciler code is manually synced to RN so you’d need to wait for a matching release.

**Comment by macrozone** [comment]
@gaearon indeed, i could reproduce the bug only on react-native.  I tried to mimick react-navigations tab navigator once in dom, once in RN:

RN (expo snack): https://snack.expo.io/H1owJ89-7
dom: https://codepen.io/anon/pen/jKxxyb

on tab change, the other sibling rerenders on RN, but on on dom

@gaearon how can i proceed? has this fix landed in RN 0.56 ?

**Comment by gaearon** [comment]
I think the fix would be in 0.56.

**Comment by jacobpdq** [comment]
I'm running 16.4.

I have two independent dom nodes, far off from each other and not children/parents of either - not even siblings in a parent.

Any update to one forces an update to another, leading me to believe this feature is a singleton for a global context. Please advise @gaearon
